README
ANoise © 2015-2020 ﻿﻿Marcos Alvarez Costales
=====================================

WHAT IS IT?
===========
An ambient noise player


HOW DO I INSTALL & RUN IT?
==========================
From a PPA:
    # add-apt-repository ppa:costales/anoise
    # apt-get update && sudo apt-get install anoise
From the code (take a look to the dependencies):
    $ bzr branch lp:anoise && cd anoise && sudo python setup.py install --prefix=/usr



DEPENDENCIES
============
python-gst0.10
gir1.2-gstreamer-0.10
gir1.2-gtk-4.0



WHAT IS NEW?
============
  * More easier to install community noises
  * Updated languages




LICENSES
====================
ANoise code is licensed under GPL v3.
http://www.gnu.org/licenses

The sound_menu.py library is licensed under GPL v3


The sounds are from the freesound.org web page, remastered by Juan Carlos (http://pragmart.wordpress.com)for ANoise project
Up to 320kb, resize length, better loop & less package size.

The sound were improved by Gordon VS http://gnome-look.org/usermanager/search.php?username=Prodigal
Fixed #1443139 Sea track has loud voices, Removed scream into Coffee shop & Better loops for Storm & Rain


App icon: CC BY 3.0 https://www.iconfinder.com/icons/352074/circle_outline_play_icon
                    https://www.iconfinder.com/icons/352468/arrow_keyboard_right_icon


Storm
    Image: CC BY 3.0 https://www.iconfinder.com/icons/390511/bolt_cloud_forecast_lightning_storm_weather_icon
    Sound: CC0 1.0   http://www.freesound.org/people/RHumphries/sounds/2523/
    
Wind
    Image: CC BY 3.0 https://www.iconfinder.com/icons/390480/eolo_forecast_weather_wind_icon
    Sound: CC0 1.0   http://www.freesound.org/people/mario1298/sounds/181252/
    
Forest
    Image: CC BY 2.5 https://www.iconfinder.com/icons/306204/forrst_icon
    Sound: CC BY 3.0 http://www.freesound.org/people/inchadney/sounds/56611/
    
Sea
    Image: CC BY 3.0 https://www.iconfinder.com/icons/216717/sea_icon
    Sound: CC BY 3.0 http://www.freesound.org/people/Owl/sounds/195294/
    
Fire
    Image: CC BY 3.0 https://www.iconfinder.com/icons/95127/fire_icon
    Sound: CC BY 3.0 http://www.freesound.org/people/juskiddink/sounds/65795/
    
Coffee Shop
    Image: CC BY 3.0 https://www.iconfinder.com/icons/111038/coffee_cup_drink_icon
    Sound: CC0 1.0   http://www.freesound.org/people/Matias44/sounds/173920/
    
Night
    Image: CC BY 3.0 https://www.iconfinder.com/icons/216526/night_weather_icon
    Sound: CC0 1.0   http://www.freesound.org/people/martypinso/sounds/22604/

Rain
    Image: CC BY 3.0 https://www.iconfinder.com/icons/390503/cloud_day_forecast_rain_shine_sun_weather_wind_icon
    Sound: CC BY 3.0 http://www.freesound.org/people/Kyster/sounds/122117/


debian/changelog
anoise (0.0.36) focal; urgency=low

  * Updated dependences

 -- costales <costales.marcos@gmail.com>  Mon, 21 apr 2020 23:47:31 +0000



anoise/utils.py
# -*- coding: utf-8 -*-
# ANoise (Ambient Noise) https://costales.github.io/projects/anoise/
# Copyright (C) 2015-2020 Marcos Alvarez Costales https://costales.github.io/
#
# ANoise is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
# 
# ANoise is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with ANoise; if not, see http://www.gnu.org/licenses
# for more information.

import os, glob, sys, socket, operator
# i18n
import gettext
gettext.textdomain('anoise')
_ = gettext.gettext


class Lock:
    """1 Instance"""
    def __init__(self):
        global lock_socket
        
        lock_socket = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
        
        try:
            lock_socket.bind('\0' + 'anoise_running') # Lock
        except socket.error:
            sys.exit() # Was locked before
        
        try:
            os.remove('/tmp/anoise_preferences')
        except:
            pass

class Noise:
    """Manage access to noises"""
    def __init__(self):
        self.CFG_DIR   = os.path.join(os.getenv('HOME'), '.config', 'anoise')
        self.CFG_FILE  = os.path.join(self.CFG_DIR, 'anoise.cfg')
        if not os.path.exists(self.CFG_DIR):
            try:
                os.makedirs(self.CFG_DIR)
            except OSError as exception:
                pass
            except:
                pass
        
        self.refresh_all_ogg()
    
    def refresh_all_ogg(self):
        """Get all current files in sounds paths"""
        all_files = []
        # Global
        sound_files = os.path.join(os.path.split(os.path.abspath(__file__))[0], 'sounds', '*.*')
        available_sounds = glob.glob(sound_files)
        for sound in available_sounds:
            if sound[-4:].lower() in '.ogg.mp3.wav':
                all_files.append(sound)
        # Local
        sound_files = os.path.join(os.getenv('HOME'), 'ANoise', '*.*')
        available_sounds = glob.glob(sound_files)
        for sound in available_sounds:
            if sound[-4:].lower() in '.ogg.mp3.wav':
                all_files.append(sound)
        sound_files = os.path.join(os.getenv('HOME'), '.ANoise', '*.*')
        available_sounds = glob.glob(sound_files)
        for sound in available_sounds:
            if sound[-4:].lower() in '.ogg.mp3.wav':
                all_files.append(sound)
        
        if not len(all_files):
            sys.exit('Not noise files found')
        
        self.noises = {}
        for noise in all_files:
            self.noises[self.get_name(noise)] = noise
        
        self.noises = sorted(self.noises.items(), key=operator.itemgetter(0))
        
        self.max = len(self.noises) - 1
        self.current = self._get_cfg_last(self.max)
        if self.current > self.max:
            self.current = 0
    
    def get_current_filename(self):
        """Get current sound filename"""
        filename = ''.join(['file://', self.noises[self.current][1]])
        return filename
    
    def set_next(self):
        """Next sound filename"""
        self.current = self.current + 1
        if self.current > self.max:
            self.current = 0
        self._set_cfg_current()
    
    def set_previous(self):
        """Previous sound filename"""
        self.current = self.current - 1
        if self.current < 0:
            self.current = self.max
        self._set_cfg_current()
    
    def get_name(self, noise=None):
        """Title for sound indicator"""
        if noise == None:
            filename = self.noises[self.current][0]
        else:
            filename = os.path.basename(noise[:-4])
            filename = filename.replace('_', ' ')
            filename = filename.title()
        return _(filename)
    
    def get_icon(self):
        """Get the name for set as Title in sound indicator"""
        filename = self.get_current_filename()[:-4]
        filename = '.'.join([filename, 'png'])
        return filename
    
    def _get_cfg_last(self, max):
        current = 0
        try:
            with open (self.CFG_FILE, "r") as myfile:
                current=int(myfile.readlines()[0])
                if current > max:
                    current = 0
        except:
            pass
        return current
    
    def _set_cfg_current(self):
        cfg_file = open(self.CFG_FILE, "w")
        cfg_file.write(str(self.current))
        cfg_file.close()


anoise/sound_menu.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
### BEGIN LICENSE
# Copyright (C) 2011 Rick Spencer <rick.spencer@canonical.com>
# This program is free software: you can redistribute it and/or modify it 
# under the terms of the GNU General Public License version 3, as published 
# by the Free Software Foundation.
# 
# This program is distributed in the hope that it will be useful, but 
# WITHOUT ANY WARRANTY; without even the implied warranties of 
# MERCHANTABILITY, SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR 
# PURPOSE.  See the GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License along 
# with this program.  If not, see <http://www.gnu.org/licenses/>.
### END LICENSE

"""Contains SoundMenuControls, A class to make it easy to integrate with the Ubuntu Sound Menu.

In order for a media player to appear in the sonud menu, it must have
a desktop file in /usr/share/applications. For example, for a media player
named "simple" player, there must be desktop file /usr/share/applications/simple-player.desktop

The desktop file must specify that it is indeed a media player. For example, simple-player.desktop
might look like the follwing:
[Desktop Entry]
Name=Simple Player
Comment=SimplePlayer application
Categories=GNOME;Audio;Music;Player;AudioVideo;
Exec=simple-player
Icon=simple-player
Terminal=false
Type=Application
MimeType=application/x-ogg;application/ogg;audio/x-vorbis+ogg;audio/x-scpls;audio/x-mp3;audio/x-mpeg;audio/mpeg;audio/x-mpegurl;audio/x-flac;

In order for the sound menu to run, a dbus loop must be running before
the player is created and before the gtk. mainloop is run. you can add
DBusGMainLoop(set_as_default=True) to your application's __main__ function.

The Ubuntu Sound Menu integrates with applications via the MPRIS2 dbus api,
which is specified here: http://www.mpris.org/2.1/spec/

This module does strive to provide an MPRIS2 implementation, but rather
focuses on the subset of functionality required by the Sound Menu.

The SoundMenuControls class can be ininstatiated, but does not provide any
default functionality. In order to provide the required functionality,
implementations must be provided for the functions starting with
"_sound_menu", such as "_sound_menu_play", etc...

Functions and properties starting with capitalize letters, such as
"Next" and "Previous" are called by the Ubuntu Sound Menu. These
functions and properties are not designed to be called directly
or overriden by application code, only the Sound Menu.

Other functions are designed to be called as needed by the 
implementation to inform the Sound Menu of changes. Thse functions
include signal_playing, signal_paused, and song_changed.

Using
#create the sound menu object and reassign functions
sound_menu = SoundMenuControls(desktop_name)
sound_menu._sound_menu_next = _sound_menu_next
sound_menu._sound_menu_previous = _sound_menu_previous
sound_menu._sound_menu_is_playing = _sound_menu_is_playing
sound_menu._sound_menu_play = _sound_menu_play
sound_menu._sound_menu_pause = _sound_menu_play
sound_menu._sound_menu_raise = _sound_menu_raise

#when the song in the player changes, it should inform
the sond menu
sound_menu.song_changed(artist,album,title)

#when the player changes to/from the playing, it should inform the sound menu
sound_menu.signal_playing()
sound_menu.signal_paused()

#whent the song is changed from the application,
#use song_changed to inform the Ubuntu Sound Menu 
sound_menu.song_changed(artist, album, song_title)

Configuring
SoundMenuControls does not come with any stock behaviors, so it 
cannot be configured

Extending
SoundMenuControls can be used as a base class with single or multiple inheritance.

_sound_menu_next
_sound_menu_previous
_sound_menu_is_playing
_sound_menu_play
_sound_menu_pause

"""

import dbus
import dbus.service

class SoundMenuControls(dbus.service.Object):
    """
    SoundMenuControls - A class to make it easy to integrate with the Ubuntu Sound Menu.

    """

    def __init__(self, desktop_name):
        """
        Creates a SoundMenuControls object.

        Requires a dbus loop to be created before the gtk mainloop,
        typically by calling DBusGMainLoop(set_as_default=True).

        arguments:
        desktop_name: The name of the desktop file for the application,
        such as, "simple-player" to refer to the file: simple-player.desktop.

        """
        self.desktop_name = desktop_name
        bus_str = """org.mpris.MediaPlayer2.%s""" % desktop_name
        bus_name = dbus.service.BusName(bus_str, bus=dbus.SessionBus())
        dbus.service.Object.__init__(self, bus_name, "/org/mpris/MediaPlayer2")
        self.__playback_status = "Stopped"

        self.song_changed()

    def song_changed(self, artists = None, album = None, title = None, album_art = None):
        """song_changed - sets the info for the current song.

        This method is not typically overriden. It should be called
        by implementations of this class when the player has changed
        songs.
            
        named arguments:
            artists - a list of strings representing the artists"
            album - a string for the name of the album
            title - a string for the title of the song

        """
        
        if artists is None:
            artists = ["Artist Unknown"]
        if album is None:
            album = "Album Uknown"
        if title is None:
            title = "Title Uknown"
        if album_art is None:
            album_art = ""
   
        self.__meta_data = dbus.Dictionary({"xesam:album":album,
                            "xesam:title":title,
                            "xesam:artist":artists,
                            "mpris:artUrl":album_art,
                            }, "sv", variant_level=1)


    @dbus.service.method('org.mpris.MediaPlayer2')
    def Raise(self):
        """Raise

        A dbus signal handler for the Raise signal. Do no override this
        function directly. rather, overrise _sound_menu_raise. This
        function is typically only called by the Sound, not directly
        from code.

        """

        self._sound_menu_raise()

    def _sound_menu_raise(self):
        """ _sound_menu_raise -

        Override this function to bring the media player to the front
        when selected by the sound menu. For example, by calling
        app_window.get_window().show()

        """

        raise NotImplementedError("""@dbus.service.method('org.mpris.MediaPlayer2') Raise
                                      is not implemented by this player.""")


    @dbus.service.method(dbus.PROPERTIES_IFACE, in_signature='ss', out_signature='v')
    def Get(self, interface, prop):
        """Get

        A function necessary to implement dbus properties.

        This function is only called by the Sound Menu, and should not
        be overriden or called directly.

        """

        my_prop = self.__getattribute__(prop)
        return my_prop

    @dbus.service.method(dbus.PROPERTIES_IFACE, in_signature='ssv')
    def Set(self, interface, prop, value):
        """Set

        A function necessary to implement dbus properties.

        This function is only called by the Sound Menu, and should not
        be overriden or called directly.

        """
        my_prop = self.__getattribute__(prop)
        my_prop = value

    @dbus.service.method(dbus.PROPERTIES_IFACE, in_signature='s', out_signature='a{sv}')
    def GetAll(self, interface):
        """GetAll

        A function necessary to implement dbus properties.

        This function is only called by the Sound Menu, and should not
        be overriden or called directly.

        """
        return {
            'CanQuit':      False,
            'CanRaise':     True,
            'HasTrackList': False,
            'DesktopEntry': self.desktop_name,
        } # Fixed #1440061

    @property
    def CanControl(self):
        '''b Read only Interface MediaPlayer2.Player'''
        return True

    @property
    def CanPlay(self):
        '''b Read only Interface MediaPlayer2.Player'''
        return True

    @property
    def CanPause(self):
        '''b Read only Interface MediaPlayer2.Player'''
        return True
    
    @property
    def CanGoNext(self):
        '''b Read only Interface MediaPlayer2.Player'''
        return True
    
    @property
    def CanGoPrevious(self):
        '''b Read only Interface MediaPlayer2.Player'''
        return True
    
    @property
    def DesktopEntry(self):
        """DesktopEntry

        The name of the desktop file.

        This propert is only used by the Sound Menu, and should not
        be overriden or called directly.

        """

        return self.desktop_name

    @property
    def PlaybackStatus(self):
        """PlaybackStatus

        Current status "Playing", "Paused", or "Stopped"

        This property is only used by the Sound Menu, and should not
        be overriden or called directly.

        """

        return self.__playback_status

    @property
    def MetaData(self):
        """MetaData

        The info for the current song.

        This property is only used by the Sound Menu, and should not
        be overriden or called directly.

        """

        return self.__meta_data

    @dbus.service.method('org.mpris.MediaPlayer2.Player')
    def Next(self):
        """Next

        A dbus signal handler for the Next signal. Do no override this
        function directly. Rather, overide _sound_menu_next. This
        function is typically only called by the Sound, not directly
        from code.

        """

        self._sound_menu_next()

    def _sound_menu_next(self): 
        """_sound_menu_next

        This function is called when the user has clicked
        the next button in the Sound Indicator. Implementations
        should overrirde this function in order to a function to
        advance to the next track. Implementations should call
        song_changed() and sound_menu.signal_playing() in order to
        keep the song information in the sound menu in sync.

        The default implementation of this function has no effect.

        """
        pass

    @dbus.service.method('org.mpris.MediaPlayer2.Player')
    def Previous(self):
        """Previous

        A dbus signal handler for the Previous signal. Do no override this
        function directly. Rather, overide _sound_menu_previous. This
        function is typically only called by the Sound Menu, not directly
        from code.

        """


        self._sound_menu_previous()

    def _sound_menu_previous(self):
        """_sound_menu_previous

        This function is called when the user has clicked
        the previous button in the Sound Indicator. Implementations
        should overrirde this function in order to a function to
        advance to the next track. Implementations should call
        song_changed() and  sound_menu.signal_playing() in order to
        keep the song information in sync.

        The default implementation of this function has no effect.


        """
        pass

    @dbus.service.method('org.mpris.MediaPlayer2.Player')
    def PlayPause(self):
        """Next

        A dbus signal handler for the Next signal. Do no override this
        function directly. Rather, overide _sound_menu_next. This
        function is typically only called by the Sound, not directly
        from code.

        """

        if not self._sound_menu_is_playing():
            self._sound_menu_play()
            self.signal_playing()
        else:
            self._sound_menu_pause()
            self.signal_paused()

    def signal_playing(self):
        """signal_playing - Tell the Sound Menu that the player has
        started playing. Implementations many need to call this function in order
        to keep the Sound Menu in synch.

        arguments:
            none

        """
        self.__playback_status = "Playing"
        d = dbus.Dictionary({"PlaybackStatus":self.__playback_status, "Metadata":self.__meta_data},
                                    "sv",variant_level=1)
        self.PropertiesChanged("org.mpris.MediaPlayer2.Player",d,[])

    def signal_paused(self):
        """signal_paused - Tell the Sound Menu that the player has
        been paused. Implementations many need to call this function in order
        to keep the Sound Menu in synch.

        arguments:
            none

        """

        self.__playback_status = "Paused"
        d = dbus.Dictionary({"PlaybackStatus":self.__playback_status},
                                    "sv",variant_level=1)
        self.PropertiesChanged("org.mpris.MediaPlayer2.Player",d,[])
            

    def _sound_menu_is_playing(self):
        """_sound_menu_is_playing         

        Check if the the player is playing,.        
        Implementations should overrirde this function 
        so that the Sound Menu can check whether to display
        Play or Pause functionality.

        The default implementation of this function always
        returns False.

        arguments:
            none

        returns:
            returns True if the player is playing, otherwise
            returns False if the player is stopped or paused.
        """

        return False

    def _sound_menu_pause(self):
        """_sound_menu_pause

        Reponds to the Sound Menu when the user has click the 
        Pause button.
        
        Implementations should overrirde this function 
        to pause playback when called.

        The default implementation of this function does nothing

        arguments:
            none

        returns:
            None
 
       """

        pass

    def _sound_menu_play(self):
        """_sound_menu_play

        Reponds to the Sound Menu when the user has click the 
        Play button.
        
        Implementations should overrirde this function 
        to play playback when called.

        The default implementation of this function does nothing

        arguments:
            none

        returns:
            None
 
       """

        pass

    @dbus.service.signal(dbus.PROPERTIES_IFACE, signature='sa{sv}as')
    def PropertiesChanged(self, interface_name, changed_properties,
                          invalidated_properties):
        """PropertiesChanged

        A function necessary to implement dbus properties.

        Typically, this function is not overriden or called directly.

        """

        pass




anoise/preferences.py
# -*- coding: utf-8 -*-
# ANoise (Ambient Noise) https://costales.github.io/projects/anoise/
# Copyright (C) 2015-2020 Marcos Alvarez Costales https://costales.github.io/
#
# ANoise is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
# 
# ANoise is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with ANoise; if not, see http://www.gnu.org/licenses
# for more information.

import gi, os, shutil, webbrowser, subprocess
from datetime import datetime, timedelta
gi.require_version('Gtk', '3.0')
gi.require_version('WebKit2', '4.0')
from gi.repository import Gtk, WebKit2
# i18n
import gettext
gettext.textdomain('anoise')
_ = gettext.gettext


class Preferences:
    """This will be for DE as MATE 14.10+ which hasn't sound indicator with Gtk3"""
    def __init__(self, player):
        self.AUTOSTART = os.path.join(os.getenv('HOME'), '.config', 'autostart', 'anoise.desktop')
        self.DESKTOP = '/usr/share/applications/anoise.desktop'
        
        self.player = player
        builder = Gtk.Builder()
        
        builder.add_from_file('/usr/share/anoise/preferences.ui')
        self.win_preferences = builder.get_object('preferences_win')
        self.adjustment   = builder.get_object('adjustment_timer')
        self.sp_timer     = builder.get_object('spin_timer')
        self.lbl_minutes  = builder.get_object('lbl_minutes')
        self.cb_sleep     = builder.get_object('cb_timesleep')
        self.cb_autostart = builder.get_object('cb_autostart')
        self.btn_noises   = builder.get_object('btn_show_noises')
        self.web          = builder.get_object('boxWeb')
        
        # Autostart
        if os.path.isfile(self.AUTOSTART):
            self.cb_autostart.set_active(True)
        else:
            self.cb_autostart.set_active(False)
        
        builder.connect_signals(self)
        
        self.win_width = self.win_height = 0
        self._save_window_size()
    
    def show(self):
        self.win_preferences.show()
    
    def _restore_window_size(self):
        self._save_window_size() # Always get the bigger widht
        self.win_preferences.set_size_request(self.win_width, self.win_height)
            
    def _save_window_size(self):
        width, height = self.win_preferences.get_size()
        if width > self.win_width:
            self.win_width = width
        if height > self.win_height:
            self.win_height = height
    
    def on_cb_autostart_toggled(self, widget, data=None):
        if self.cb_autostart.get_active():
            try:
                shutil.copy2(self.DESKTOP, self.AUTOSTART)
            except:
                pass
        else:
            try:
                os.remove(self.AUTOSTART)
            except:
                pass
    
    def set_show_timer(self):
        if self.cb_sleep.get_active():
            self.cb_sleep.set_active(False)
    
    def on_cb_timesleep_toggled(self, widget, data=None):
        seconds = self.sp_timer.get_value_as_int() * 60
        self.sp_timer.set_sensitive(not self.cb_sleep.get_active())
        self.player.set_timer(self.cb_sleep.get_active(), seconds)
        if self.cb_sleep.get_active():
            self.lbl_minutes.hide()
            self.sp_timer.hide()
            x = datetime.now() + timedelta(seconds=seconds)
            msg = ' '.join([_("ANoise will stop at"), x.strftime('%H:%M')])
            self.cb_sleep.set_label(msg)
            self._restore_window_size()
        else:
            self.lbl_minutes.show()
            self.sp_timer.show()
            self.cb_sleep.set_label(_("Stop in"))
            self._restore_window_size()
    
    def on_btn_show_noises_clicked(self, widget, data=None):
        self.btn_noises.hide()
        web_content = WebKit.WebView()
        settings = web_content.get_settings()
        settings.set_property('enable-default-context-menu', False)
        web_content.set_settings(settings)
        web_content.open('https://costales.github.io/projects/anoise/')
        web_content.connect('navigation-requested', self._on_navigation_requested)
        self.web.add(web_content)
        self.web.show_all()
    
    def _on_navigation_requested(self, view, frame, req):
        uri = req.get_uri()
        if uri and uri.startswith('apt'): # OS installer
            subprocess.Popen(["apturl", uri], shell=False)
        if uri and uri.startswith('http'): # authors
            webbrowser.open(uri)
        return True
    
    def on_preferences_delete_event(self, widget, data=None):
        self.win_preferences.hide()
        return True


anoise/anoise.py
# -*- coding: utf-8 -*-
# ANoise 0.0.34 (Ambient Noise) https://costales.github.io/projects/anoise/
# Copyright (C) 2015-2020 Marcos Alvarez Costales https://costales.github.io/
#
# ANoise is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
# 
# ANoise is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with ANoise; if not, see http://www.gnu.org/licenses
# for more information.

import gi, os, threading
gi.require_version('Gtk', '3.0')
gi.require_version('Gst', '1.0')
from gi.repository import Gtk, GObject, Gst
from dbus.mainloop.glib import DBusGMainLoop
from utils import *
from sound_menu import SoundMenuControls
from preferences import Preferences
try:
    from view import GUI
except ImportError:
    pass

# i18n
import gettext
gettext.textdomain('anoise')
_ = gettext.gettext

# playbin breaks in Kubuntu 14.04 > Needs Gst 0.10
try:
    gi.require_version('Gst', '1.0')
except:
    gi.require_version('Gst', '0.10')
    PLAYBIN = "playbin2"
else:
    PLAYBIN = "playbin"


class ANoise:
    """Control the sound indicator"""
    def __init__(self):
        DBusGMainLoop(set_as_default=True)
        Gst.init(None)
        
        self.sound_menu = SoundMenuControls('anoise')
        self.noise = Noise()
        self.win_preferences = Preferences(self)
        
        # Need in a few DE
        try:
            self.window = GUI(self)
        except:
            pass
        
        self.player = Gst.ElementFactory.make(PLAYBIN, "player")
        self.player.connect("about-to-finish", self._loop)
        
        self.player.set_property('uri', self.noise.get_current_filename())
        self.is_playing = True
        
        dummy_i18n = (_("Coffee Shop"), _("Fire"), _("Forest"), _("Night"), _("Rain"), _("River"), _("Sea"), _("Storm"), _("Wind")) # Need i18n
        
        # Overwrite libraty methods
        self.sound_menu._sound_menu_is_playing = self._sound_menu_is_playing
        self.sound_menu._sound_menu_play       = self._sound_menu_play
        self.sound_menu._sound_menu_pause      = self._sound_menu_pause
        self.sound_menu._sound_menu_next       = self._sound_menu_next
        self.sound_menu._sound_menu_previous   = self._sound_menu_previous
        self.sound_menu._sound_menu_raise      = self._sound_menu_raise
        
        # Autostart when click on sound indicator icon
        threading.Timer(2, self._sound_menu_play).start()
    
    def _loop(self, message):
        """Start again the same sound in the EOS"""
        self.player.set_property('uri', self.noise.get_current_filename())
    
    def _sound_menu_is_playing(self):
        """Called in the first click"""
        return self.is_playing
    
    def _sound_menu_play(self):
        """Play"""
        self.is_playing = True # Need to overwrite this for an issue with autstart
        self.sound_menu.song_changed('', '', self.noise.get_name(), self.noise.get_icon())
        self.player.set_state(Gst.State.PLAYING)
        self.sound_menu.signal_playing()
    
    def _sound_menu_pause(self):
        """Pause"""
        self.is_playing = False # Need to overwrite this for an issue with autstart
        self.player.set_state(Gst.State.PAUSED)
        self.sound_menu.signal_paused()
    
    def _set_new_play(self, what):
        """Next or Previous"""
        self.noise.refresh_all_ogg()
        # Get Next/Previous
        if what == 'next':
            self.noise.set_next()
        if what == 'previous':
            self.noise.set_previous()
        # From pause?
        self.player.set_state(Gst.State.READY)
        if not self.is_playing:
            self.is_playing = True
        # Set new sound
        self.player.set_property('uri', self.noise.get_current_filename())
        # Play
        self._sound_menu_play()
    
    def _sound_menu_previous(self):
        """Previous"""
        self._set_new_play('previous')
    
    def _sound_menu_next(self):
        """Next"""
        self._set_new_play('next')
    
    def _sound_menu_raise(self):
        """Click on player"""
        self.win_preferences.show()
    
    def set_timer(self, enable, seconds):
        if enable:
            self.timer = threading.Timer(seconds, self._set_future_pause)
            self.timer.start()
        else:
            self.timer.cancel()
    
    def _set_future_pause(self):
        self.win_preferences.set_show_timer()
        self._sound_menu_pause()
    

if __name__ == "__main__":
    Lock()
    anoise = ANoise()
    Gtk.main()
